cupy/_core/core.pyx

```diff
@@ -342,6 +354,44 @@ cdef class _ndarray_base:
             device_type = dlpack.device_ROCM
         return (device_type, self.device.id)
 
+    def __getbuffer__(self, Py_buffer* buf, int flags):
+        # TODO(leofang): use flags
+        if (not is_sam_supported(self.data.device_id)
+                or not self.is_host_accessible()):
+            raise RuntimeError(
+                'Accessing a CuPy ndarry on CPU is not allowed except when '
+                'using system memory (on HMM or ATS enabled systems, need to '
+                'set CUPY_ENABLE_SAM=1) or managed memory')
+
+        populate_format(buf, self.dtype.char)
+        buf.buf = <void*><intptr_t>self.data.ptr
+        buf.itemsize = self.dtype.itemsize
+        buf.len = self.size
+        buf.internal = NULL
+        buf.readonly = 0  # TODO(leofang): use flags
+        cdef int n, ndim
+        ndim = self._shape.size()
+        cdef Py_ssize_t* shape_strides = <Py_ssize_t*>stdlib.malloc(
+            sizeof(Py_ssize_t) * ndim * 2)
+        for n in range(ndim):
+            shape_strides[n] = self._shape[n]
+            shape_strides[n + ndim] = self._strides[n]  # in bytes
+        buf.ndim = ndim
+        buf.shape = shape_strides
+        buf.strides = shape_strides + ndim
+        buf.suboffsets = NULL
+        buf.obj = self
+        cpython.Py_INCREF(self)
+
+        stream_module.get_current_stream().synchronize()
+
+    def __releasebuffer__(self, Py_buffer* buf):
+        stdlib.free(buf.shape)  # frees both shape & strides
+        cpython.Py_DECREF(self)
+
+    cdef inline bint is_host_accessible(self) except*:
+        return self.data.mem.identity in ('SystemMemory', 'ManagedMemory')
```

Comment:
Note: This is where we have a strict requirement for what kinds of memories that can be used for unified np/cp programming